### ISSUE TITLE
"IDE0004 C# Cast is redundant" is reported incorrectly. Strange fix behavior.

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#20211]
**Version Used**: 
VisualStudio.15.Preview/15.3.0-Preview+26510.0

**Steps to Reproduce**:

Consider the following program:

```cs
class Program
{
  public static void Main()
  {
    switch ((object)null) // here
    {
      case bool _:
      case true:
      case false:
        break;
    }

    switch ((bool?)null) // here
    {
      case bool _:
      case true:
      case false:
        break;
    }
  }
}
```

**Expected Behavior**:

"IDE0004 C# Cast is redundant" is not reported. The program obviously won't compile without the casts.

**Actual Behavior**:
Both casts are highlighted as redundant by "IDE0004 C# Cast is redundant".
Fix 'Remove unnecessary cast' is offered but does nothing.
However I can select 'preview changes' from the fix menu and apply it from the preview window. It will remove the cast and broke compilation although the preview window says there will be no changes.

![ide0004](https://user-images.githubusercontent.com/5566206/27135887-69bb383c-5122-11e7-9f7d-9c7460f84458.gif)

**Comment by Pilchie** [comment]
Looks like a duplicate of #19987.  @sharwell do you concur?

**Comment by sharwell** [comment]
@Pilchie One part of this might be a duplicate. The weird preview/apply behavior is also concerning though.

**Comment by sharwell** [comment]
It may turn out that this is a combined duplicate of several other items. I'll investigate to make sure nothing here is new.
